tsdb/importer/import.go

```diff
@@ -0,0 +1,310 @@
+		}
+
+		var lset prom_labels.Labels
+		_ = parser.Metric(&lset)
+
+		sample := &tsdb.MetricSample{TimestampMs: currentTimestampMs, Value: val, Labels: tsdb_labels.FromMap(lset.Map())}
+
+		// Run this flow when current sample is outside the DB time limits
+		// OR
+		// when there are no blocks in the DB.
+		if sample.TimestampMs < dbMint || sample.TimestampMs > dbMaxt || len(blockTimes) == 0 || blockTimes == nil {
+			// In the case where we are importing data from before the DB
+			// start time, we could run into a case where the imported data
+			// could overlap with current data. This flag sets up the proper
+			// flow for that case.
+			if sample.TimestampMs < dbMint {
+				haveSamplesBeforeDbStart = true
+			}
+			// Take care of samples that are on either side of the DB's time limits.
+			if len(currentBlock) > 0 {
+				startTime, endTime := currentBlock[0].TimestampMs, currentBlock[len(currentBlock)-1].TimestampMs
+				// Check if adding current sample will push current block over 2h, if so, cut and push
+				// OR
+				// If we already have samples in the case where the current sample just goes past the
+				// DB maxt - then we must push the current blocks to disk, before accepting the new one.
+				if (sample.TimestampMs-startTime) >= tsdb.DefaultBlockDuration ||
+					haveSamplesInsideBlockTimeRange {
+					var parts [][]*tsdb.MetricSample
+					if haveSamplesInsideBlockTimeRange {
+						parts = partitionBlocks(currentBlock, endTime-startTime)
+					} else {
+						parts = partitionBlocks(currentBlock, tsdb.DefaultBlockDuration)
+					}
+					for _, part := range parts {
+						path, err := tsdb.CreateBlock(part, dbPath, logger)
+						if err != nil {
+							return blockPaths, err
+						}
+						blockPaths = append(blockPaths, path)
+					}
+					currentBlock = []*tsdb.MetricSample{}
+					if haveSamplesInsideBlockTimeRange {
+						haveSamplesInsideBlockTimeRange = false
+					}
+				}
+			}
+			currentBlock = append(currentBlock, sample)
+			continue
+		}
+		// This will only be triggered in the case where we have some
+		// samples from before the DB mint, that also overlap with the
+		// DB blocks times.
+		if haveSamplesBeforeDbStart {
+			parts := partitionBlocks(currentBlock, tsdb.DefaultBlockDuration)
+			for _, part := range parts {
+				path, err := tsdb.CreateBlock(part, dbPath, logger)
+				if err != nil {
+					return blockPaths, err
+				}
+				blockPaths = append(blockPaths, path)
+			}
+			currentBlock = []*tsdb.MetricSample{}
+			// Once this is set to false in here, it will never be triggered again,
+			// as we have already crossed into the DB's time limits now.
+			haveSamplesBeforeDbStart = false
+		}
+		if blockTimes[blockIdx].start <= sample.TimestampMs && sample.TimestampMs <= blockTimes[blockIdx].end {
+			currentBlock = append(currentBlock, sample)
+			haveSamplesInsideBlockTimeRange = true
+		} else {
+			if len(currentBlock) > 0 {
+				// Push to disk, no need to partition.
```

Comment:
I'm not sure that's true, what if it runs into the head block?

Reply:
Can you give me an example of such a case occurring so that I can test it out? Because I don't quite understand what you mean.

Reply:
Data from the last hour